lib/aliquot/validator.rb

```diff
@@ -166,59 +160,95 @@ class SignedMessageContract < Dry::Validation::Contract
-          !values[:cryptogram].nil?
+      rule(:pan).validate(:integer_string?, :pan?)
+    end
+
+    class ECv1_TokenizedPaymentMethodDetailsContract < Dry::Validation::Contract
+      json(CommonPaymentMethodDetailsContract.schema) do
+        required(:dpan).filled(:str?)
+        required(:threedsCryptogram).filled(:str?)
+        required(:eciIndicator).filled(:str?)
+        required(:authMethod).filled(:str?, included_in?: %w[3DS])
       end
 
-      rule(:eciIndicator) do
-        key.failure('cannot be defined') if 'PAN_ONLY'.eql?(values[:authMethod]) &&
-          !values[:eciIndicator].nil?
+      rule(:dpan).validate(:integer_string?, :pan?)
+      rule(:eciIndicator).validate(:eci?)
+    end
+
+    class ECv2_PaymentMethodDetailsContract < Dry::Validation::Contract
+      json(CommonPaymentMethodDetailsContract.schema) do
+        required(:pan).filled(:str?)
+        required(:authMethod).filled(:str?, included_in?: %w[PAN_ONLY])
       end
+
+      rule(:pan).validate(:integer_string?, :pan?)
     end
 
-    PaymentMethodDetailsSchema = PaymentMethodDetailsContract.new
+    class ECv2_TokenizedPaymentMethodDetailsContract < Dry::Validation::Contract
+      json(CommonPaymentMethodDetailsContract.schema) do
+        required(:pan).filled(:str?)
+        required(:cryptogram).filled(:str?)
+        required(:eciIndicator).filled(:str?)
+        required(:authMethod).filled(:str?, included_in?: %w[CRYPTOGRAM_3DS])
+      end
+
+      rule(:pan).validate(:integer_string?, :pan?)
+      rule(:eciIndicator).validate(:eci?)
+    end
+
+    # PaymentMethodDetailsSchema = PaymentMethodDetailsContract.new
 
     # DRY-Validation schema for encryptedMessage component Google Pay token
     class EncryptedMessageContract < Dry::Validation::Contract
       json do
         required(:messageExpiration).filled(:str?)
         required(:messageId).filled(:str?)
         required(:paymentMethod).filled(:str?)
-        required(:paymentMethodDetails).filled(:hash).schema(PaymentMethodDetailsContract.schema)
+        required(:paymentMethodDetails).filled(:hash)
         optional(:gatewayMerchantId).filled(:str?)
       end
       rule(:messageExpiration).validate(:integer_string?)
+
+      rule(:paymentMethodDetails).validate do
+        if values[:protocolVersion] == 'ECv1'
+          if values[:paymentMethod] == 'TOKENIZED_CARD'
+            return ECv1_TokenizedPaymentMethodDetailsContract.new
+          else
+            return ECv1_PaymentMethodDetailsContract.new
+          end
+        else
+          if  values[:authMethod] == 'CRYPTOGRAM_3DS'
+            return ECv2_TokenizedPaymentMethodDetailsContract.new
+          else
+            return ECv2_PaymentMethodDetailsContract.new
+          end
+        end
+      end
       rule(:paymentMethod) do
-        key.failure('must be equal to CARD') unless 'CARD'.eql?(value)
+        if values[:paymentMethodDetails] && values[:paymentMethodDetails].is_a?(Hash)
```

Comment:
`values[:paymentMethodDetails] &&` can be removed as it's covered by `is_a?(Hash)`, but no biggie if you prefer to be explicit.